papers/yellow/voting.tex

```diff
@@ -69,10 +70,61 @@ \subsection{Proposing a change and voting for it}
 
 \subsection{Voting for a soft-fork}
 
-Soft-fork is when a protocol version supported by the network is being increased. This version is written in a block
-header.
+A soft-fork happens when a protocol version supported by the network is being increased. This version is written in a block
+header. Semantics behind versioning is not defined ahead of time by the protocol and so up to clients and their
+developers. The protocol version is being written into every block header, with initial value during launch to be set
+to 1.
 
-To start voting for a soft-fork, a miner needs to publish identifier $120$ in the first block of the epoch, consider,
+The protocol version upgrade is to be done in the following steps:
+\begin{enumerate}
+    \item{} A miner is proposing to increase protocol version.
+    \item{} Other miners are voting within 32 epochs for the proposal
+    \item{} If the soft-fork proposal is being rejected~(so if it gathers no more than $90\%$ of votes during voting period, e.i.
+            no more than $\lfloor 32 * 1024 * 90 / 100 \rfloor = 29491$ votes), new voting may be proposed next epoch
+            after the voting done.
+    \item{} If the soft-fork proposal is being approved, activation period of 128 epochs is starting. The first block
+    after the activation period is called activation height. It is prohibited to vote for a new soft-fork during the
+    activation and also on the activation height. Soft-fork data is still to be
+    written to corresponding extension sections~(see below) during the activation period, and on activation height.
+    Block version written into extensions is to be increased from the first block of the activation period, while protocol
+    version in headers is still the same. Protocol version in headers is increased from the activation height.
+\end{enumerate}
+
+
+To start voting for a soft-fork, a miner needs to publish the identifier $120$ in the first block of the epoch, consider,
 for example, that its height is $h_s$. Next epoch, a miner should post height when start fork was proposed ($122: h_s$)
 in the extension section of the block, and number of votes collected in the previous epochs $v_s$ should be written
-there as well as $(121, v_s)$.
\ No newline at end of file
+there as well as ($121: v_s$).
+
+Examples:
+
+\begin{itemize}
+    \item{} Assume that a vote for soft-fork is proposed in block number $1024$ (by writing a record into the extension
```

Comment:
Do we have tests for all these rules?
Let's write tests for all the rules above, e.g. test that checks that:
- during soft-fork voting  extensions with incorrect values in `SoftForkStartingHeight` and `SoftForkVotesCollected` will be rejected
- during soft-fork voting extensions with missed `SoftForkStartingHeight` and `SoftForkVotesCollected` keys will be rejected
- extensions that contain `SoftForkStartingHeight` and `SoftForkVotesCollected` keys if soft-fork working is not started will be rejected
- an extension that does not clear voting parameters after unsuccessful voting will be rejected
- an extension that does not contain voting parameters during the activation period will be rejected
- an extension that does not increase block version in extension after the activation period will be rejected
- and so on